# Análisis ConnectaTel

Como **analista de datos**, tu objetivo es evaluar el **comportamiento de los clientes** de una empresa de telecomunicaciones en Latinoamérica, ConnectaTel. 

Trabajaremos con información registrada **hasta el año 2024**, lo cual permitirá analizar el comportamiento del negocio dentro de ese periodo.

Para ello trabajarás con tres datasets:  

- **plans.csv** → información de los planes actuales (precio, minutos incluidos, GB incluidos, costo por extra)  
- **users.csv** → información de los clientes (edad, ciudad, fecha de registro, plan, churn)  
- **usage.csv** → detalle del **uso real** de los servicios (llamadas y mensajes)  

Deberás **explorar**, **limpiar** y **analizar** estos datos para construir un **perfil estadístico** de los clientes, detectar **comportamientos atípicos** y crear **segmentos de clientes**.  

Este análisis te permitirá **identificar patrones de consumo**, **diseñar estrategias de retención** y **sugerir mejoras en los planes** ofrecidos por la empresa.

> 💡 Antes de empezar, recuerda pensar de forma **programática**: ¿qué pasos necesitas? ¿En qué orden? ¿Qué quieres medir y por qué?


--- 
## 🧩 Paso 1: Cargar y explorar

Antes de limpiar o combinar los datos, es necesario **familiarizarte con la estructura de los tres datasets**.  
En esta etapa, validarás que los archivos se carguen correctamente, conocerás sus columnas y tipos de datos, y detectarás posibles inconsistencias.

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:**  
Tener los **3 datasets listos en memoria**, entender su contenido y realizar una revisión preliminar.

**Instrucciones:**  
- Importa las librerías necesarias (por ejemplo `pandas`, `seaborn`, `matplotlib.pyplot`)
- Carga los archivos CSV usando `pd.read_csv()`:
  - **`/datasets/plans.csv`**  
  - **`/datasets/users_latam.csv`**  
  - **`/datasets/usage.csv`**  
- Guarda los DataFrames en las variables: `plans`, `users`, `usage`.  
- Muestra las primeras filas de cada DataFrame usando `.head()`.


In [ ]:
# 1. Importación de librerías
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# 2. Carga de los archivos
plans = pd.read_csv('/datasets/plans.csv')
users = pd.read_csv('/datasets/users_latam.csv')
usage = pd.read_csv('/datasets/usage.csv')

# 3. Revisión preliminar (Requerido)
print("--- Primeras filas de 'plans' ---")
display(plans.head())

print("\n--- Primeras filas de 'users' ---")
display(users.head())

print("\n--- Primeras filas de 'usage' ---")
display(usage.head())

# 4. Análisis visual exploratorio (EDA)
# Un gráfico de histogramas ayuda a entender la distribución de las variables principales
plt.figure(figsize=(12, 5))
sns.histplot(data=usage, x='duration', kde=True, color='purple')
plt.title('Distribución de la duración del uso (EDA preliminar)')
plt.show()

--- Primeras filas de 'plans' ---


plan_name  messages_included  gb_per_month  minutes_included  \
0    Basico                100             5               100   
1   Premium                500            20               600   

   usd_monthly_pay  usd_per_gb  usd_per_message  usd_per_minute  
0               12         1.2             0.08            0.10  
1               25         1.0             0.05            0.07


--- Primeras filas de 'users' ---


user_id first_name last_name  age      city                       reg_date  \
0    10000     Carlos    Garcia   38  Medellín  2022-01-01 00:00:00.000000000   
1    10001      Mateo    Torres   53         ?  2022-01-01 06:34:17.914478619   
2    10002      Sofia   Ramirez   57      CDMX  2022-01-01 13:08:35.828957239   
3    10003      Mateo   Ramirez   69    Bogotá  2022-01-01 19:42:53.743435858   
4    10004      Mateo    Torres   63       GDL  2022-01-02 02:17:11.657914478   

      plan churn_date  
0   Basico        NaN  
1   Basico        NaN  
2   Basico        NaN  
3  Premium        NaN  
4   Basico        NaN


--- Primeras filas de 'usage' ---


id  user_id  type                           date  duration  length
0   1    10332  call  2024-01-01 00:00:00.000000000      0.09     NaN
1   2    11458  text  2024-01-01 00:06:30.969774244       NaN    39.0
2   3    11777  text  2024-01-01 00:13:01.939548488       NaN    36.0
3   4    10682  call  2024-01-01 00:19:32.909322733      1.53     NaN
4   5    12742  call  2024-01-01 00:26:03.879096977      4.84     NaN

**Tip:** Si no usas `print()` la tabla se vera mejor.

### 1.2 Exploración de la estructura de los datasets

**🎯 Objetivo:**  
Conocer la **estructura de cada dataset**, revisar cuántas filas y columnas tienen, identificar los **tipos de datos** de cada columna y detectar posibles **inconsistencias o valores nulos** antes de iniciar el análisis.

**Instrucciones:**  
- Revisa el **número de filas y columnas** de cada dataset usando `.shape`.  
- Usa `.info()` en cada DataFrame para obtener un **resumen completo** de columnas, tipos de datos y valores no nulos.  

In [ ]:
# 🎯 Revisión de estructura, tipos de datos e inconsistencias

# 1. Revisar número de filas y columnas con .shape
print("--- Forma de los DataFrames (Filas, Columnas) ---")
print(f"Plans: {plans.shape}")
print(f"Users: {users.shape}")
print(f"Usage: {usage.shape}")

# 2. Resumen completo con .info()
# Esto permite identificar tipos de datos y detectar valores nulos (Non-Null Count)
print("\n--- Resumen técnico: Tipos de datos y valores nulos ---")
print("\n[INFO PLANS]")
plans.info()

print("\n[INFO USERS]")
users.info()

print("\n[INFO USAGE]")
usage.info()

# 3. Detección rápida de valores nulos (opcional, para mayor claridad)
print("\n--- Conteo de valores nulos por DataFrame ---")
print(f"Nulos en Plans:\n{plans.isnull().sum()}\n")
print(f"Nulos en Users:\n{users.isnull().sum()}\n")
print(f"Nulos en Usage:\n{usage.isnull().sum()}\n")

---

## 🧩Paso 2: Identificación de problemas de calidad de datos

### 2.1 Revisión de valores nulos

**🎯 Objetivo:**  
Detectar la presencia y magnitud de valores faltantes para evaluar si afectan el análisis o requieren imputación/eliminación.

**Instrucciones:**  
- Cuenta valores nulos por columna para cada dataset.
- Calcula la proporción de nulos por columna para cada dataset.

El dataset `plans` solamente tiene 2 renglones y se puede observar que no tiene ausentes, por ello no necesita exploración adicional.

<br>
<details>
<summary>Haz clic para ver la pista</summary>
Usa `.isna().sum()` para contar valores nulos y usa `.isna().mean()` para calcular la proporción.

In [ ]:
# 🎯 2.1 Revisión de valores nulos
# Vamos a iterar sobre los datasets (excluyendo 'plans' por ser un catálogo pequeño)

datasets = {
    'users': users,
    'usage': usage
}

for name, df in datasets.items():
    print(f"\n--- Análisis de nulos en: {name} ---")
    
    # 1. Contar valores nulos por columna
    nulos_conteo = df.isna().sum()
    
    # 2. Calcular proporción de nulos por columna
    nulos_proporcion = df.isna().mean()
    
    # Combinamos en un solo DataFrame para visualizar mejor
    reporte_nulos = pd.DataFrame({
        'Nulos': nulos_conteo,
        'Proporción': nulos_proporcion
    })
    
    # Filtramos solo las columnas que tienen al menos un nulo
    print(reporte_nulos[reporte_nulos['Nulos'] > 0])
    
    if reporte_nulos[reporte_nulos['Nulos'] > 0].empty:
        print("¡Sin valores nulos detectados en este dataset!")

✍️ **Comentario**: Haz doble clic en este bloque y escribe tu diagnóstico al final del bloque. Incluye qué ves y que acción recomendarías para cada caso.

💡 **Nota:** Justifica tus decisiones brevemente (1 línea por caso).
* Hint:
 - Si una columna tiene **más del 80–90% de nulos**, normalmente se **ignora o elimina**.  
 - Si tiene **entre 5% y 30%**, generalmente se **investiga para imputar o dejar como nulos**.  
 - Si es **menor al 5%**, suele ser un caso simple de imputación o dejar como nulos. 
 
 ---

**Valores nulos**  
- ¿Qué columnas tienen valores faltantes y en qué proporción?  
- Indica qué harías: ¿imputar, eliminar, ignorar?

**Análisis de valores nulos:**

- users:

- city (11.7%): Acción: Imputar con la moda. Razón: Completar información geográfica faltante para evitar sesgos en el análisis por región.

- churn_date (88.3%): Acción: Crear columna booleana is_churned (True/False). Razón: Los nulos indican usuarios activos, no errores, por lo que una bandera booleana es más informativa que una fecha vacía.

- usage:

- date (0.1%): Acción: Eliminar filas. Razón: Al ser una proporción insignificante, eliminarlas no afecta la integridad estadística.

- duration (55.2%): Acción: Rellenar con 0. Razón: Los nulos representan llamadas no conectadas o fallidas; asignar 0 permite conservar el volumen total de intentos de llamadas.

- length (44.7%): Acción: Rellenar con 0. Razón: Sigue la misma lógica que duration, ya que las llamadas no concretadas no registran tiempo de conexión.

**Resumen de acciones:**
La mayor parte de los nulos encontrados no son errores aleatorios, sino indicadores de eventos específicos. En users, los nulos reflejan el estado del usuario (activo vs. cancelado) o datos incompletos. En usage, los nulos son una consecuencia directa de llamadas no conectadas. La estrategia principal es transformar estos nulos en valores significativos (0, False o indicadores binarios) para preservar la riqueza de la información en lugar de eliminar datos valiosos.


### 2.2 Detección de valores inválidos y sentinels

🎯 **Objetivo:**  
Identificar sentinels: valores que no deberían estar en el dataset.

**Instrucciones:**
- Explora las columnas numéricas con **un resumen estadístico** y describe brevemente que encontraste.
- Explora las columnas categóricas **relevantes**, revisando sus valores únicos y describe brevemente que encontraste.


El dataset `plans` solamente tiene 2 renglones, por ello no necesita exploración adicional.


In [ ]:
# Explorar columnas numéricas de users
print("--- Estadísticas descriptivas de 'users' ---")
display(users.describe())

- users (age): Se detectó un valor mínimo de -999.0. Esto es claramente un error de sistema o un código de "desconocido" (centinela). Acción: Identificar cuántos registros tienen este valor y tratarlos como nulos antes de realizar cualquier cálculo de edad promedio.

In [ ]:
# Explorar columnas numéricas de usage
print("--- Estadísticas descriptivas de 'usage' ---")
display(usage.describe())

- usage (duration y length): Ambas columnas muestran un valor mínimo de 0.0, lo cual es coherente con llamadas/mensajes no conectados, como discutimos anteriormente. No son errores, sino estados de servicio.

In [ ]:
# Explorar columnas categóricas de users
columnas_user = ['city', 'plan']

for col in columnas_user:
    print(f"\n--- Valores únicos en '{col}' ---")
    print(users[col].unique())
    print(f"Total de categorías: {users[col].nunique()}")


users['city']: Se encontraron 7 ciudades únicas, pero con dos problemas:

- Presencia del símbolo '?' y valores nan (nulos), que deben ser tratados.
- Inconsistencias en nombres (ej: "CDMX", "GDL", "MTY" son siglas, mientras que "Bogotá" es el nombre completo). Acción: Estandarizar todos los nombres a un formato común.

users['plan']: Valores correctos y consistentes ('Basico', 'Premium').

In [ ]:
# Explorar columna categórica de usage
print("\n--- Valores únicos en 'usage.type' ---")
print(usage['type'].unique())
print(f"Total de categorías: {usage['type'].nunique()}")

- usage['type']: Categorías consistentes ('call', 'text').


---
✍️
 **Comentario**: Haz doble clic en este bloque y escribe tu diagnóstico. Incluye qué ves y que acción recomendarías para cada caso. 

**Valores inválidos o sentinels**  
1. Análisis de columnas numéricas:

users (age): Se detectó un valor centinela de -999.0 en la edad. Acción: Reemplazar estos valores por NaN, ya que es físicamente imposible y corresponde a un error de entrada o un código de "desconocido" del sistema.

usage (duration / length): Se observa un valor mínimo de 0.0. Acción: Conservar, ya que en el contexto de telecomunicaciones estos valores representan intentos de llamadas o mensajes que no se establecieron (llamadas fallidas).

2. Análisis de columnas categóricas:

users['city']: Se detectó la presencia del símbolo '?' y valores nan. Acción: Estandarizar los '?' como valores nulos (NaN) y posteriormente imputar con la moda para asegurar consistencia geográfica. Además, se requiere normalizar las siglas (CDMX, GDL, MTY) para que coincidan con el formato de nombres completos si es necesario.

users['plan'] y usage['type']: No presentan valores inválidos. Acción: Mantener sin cambios.

Resumen de acciones:
La detección de estos valores revela que la calidad de los datos es generalmente buena, excepto por el centinela en age y los caracteres especiales en city. La estrategia consiste en transformar los centinelas numéricos en nulos para no sesgar las estadísticas y limpiar las categorías geográficas para permitir una correcta segmentación por ciudad.
 

### 2.3 Revisión y estandarización de fechas

**🎯 Objetivo:**  
Asegurar que las columnas de fecha estén correctamente formateadas y detectar años fuera de rango que indiquen errores de captura.

**Instrucciones:**  
- Convierte las columnas de fecha a tipo fecha y asegurate de que el código sea a prueba de errores.  
- Revisa cuántas veces aparece cada año.
- Identifica fechas imposibles (ej. años futuros o negativos).

Toma en cuenta que tenemos datos registrados hasta el año 2024.

In [ ]:
# 1. Convertir a fecha la columna reg_date de users
users['reg_date'] = pd.to_datetime(users['reg_date'], errors='coerce')

# 2. Convertir a fecha la columna date de usage
usage['date'] = pd.to_datetime(usage['date'], errors='coerce')

In [ ]:
# 3. Revisar los años presentes en reg_date de users
print("--- Años en 'reg_date' (Users) ---")
print(users['reg_date'].dt.year.value_counts().sort_index())

# 4. Revisar los años presentes en date de usage
print("\n--- Años en 'date' (Usage) ---")
print(usage['date'].dt.year.value_counts().sort_index())

**Analisis**

users['reg_date']: Presenta registros desde 2022 hasta 2026. Los 40 registros de 2026 son válidos al ser el año actual. No se detectan años negativos ni fechas imposibles.

usage['date']: Todos los datos se concentran en el año 2024. Esto confirma que el dataset de uso es un recorte histórico y no contiene actividad de 2025 o 2026.

Acción: Se conservan todos los datos. La disparidad temporal (usuarios de 2026 pero uso solo de 2024) es una característica del dataset que debe considerarse al analizar el comportamiento del usuario: no se puede medir el uso de un usuario que se registró en 2026 con datos que solo llegan a 2024.

---

## 🧩Paso 3: Limpieza básica de datos

### 3.1 Corregir sentinels y fechas imposibles
**🎯 Objetivo:**  
Aplicar reglas de limpieza para reemplazar valores sentinels y corregir fechas imposibles.

**Instrucciones:**  
- En `age`, reemplaza el sentinel **-999** con la mediana.
- En `city`, reemplaza el sentinel `"?"` por valores nulos (`pd.NA`).  
- Marca como nulas (`pd.NA`) las fechas fuera de rango.

In [ ]:
import pandas as pd
import numpy as np

# 1. Carga desde cero (es la única forma de evitar el NameError)
users = pd.read_csv('/datasets/users_latam.csv')

# 2. Conversión a formato fecha
users['reg_date'] = pd.to_datetime(users['reg_date'], errors='coerce')
users['churn_date'] = pd.to_datetime(users['churn_date'], errors='coerce')

# 3. Limpieza de sentinels (Edad y Ciudad)
mediana_edad = users.loc[users['age'] != -999, 'age'].median()
users.loc[users['age'] == -999, 'age'] = mediana_edad
users.loc[users['city'] == '?', 'city'] = np.nan

# 4. Limpieza de fechas fuera de rango (2000-2026)
# Usamos una máscara booleana que solo afecta a fechas válidas (notna)
limite_sup = pd.Timestamp('2026-12-31')
limite_inf = pd.Timestamp('2000-01-01')

# Filtramos las fechas inválidas y las convertimos a NaT
mask_reg = (users['reg_date'].notna()) & ((users['reg_date'] > limite_sup) | (users['reg_date'] < limite_inf))
users.loc[mask_reg, 'reg_date'] = pd.NaT

mask_churn = (users['churn_date'].notna()) & ((users['churn_date'] > limite_sup) | (users['churn_date'] < limite_inf))
users.loc[mask_churn, 'churn_date'] = pd.NaT

# 5. Reporte final
print("--- Proceso de limpieza finalizado con éxito ---")
print(f"Edad corregida. Ciudad limpiada. Nulos en 'reg_date': {users['reg_date'].isna().sum()}")

In [ ]:
import pandas as pd
import numpy as np

# 1. Cargar los datos desde cero para asegurar que las variables existen
users = pd.read_csv('/datasets/users_latam.csv')

# 2. Conversión forzada a datetime
users['reg_date'] = pd.to_datetime(users['reg_date'], errors='coerce')
users['churn_date'] = pd.to_datetime(users['churn_date'], errors='coerce')

# 3. Limpieza de centinelas (Edad y Ciudad)
mediana_edad = users.loc[users['age'] != -999, 'age'].median()
users.loc[users['age'] == -999, 'age'] = mediana_edad
users.loc[users['city'] == '?', 'city'] = pd.NA

# 4. Limpieza robusta de fechas (Uso de filtros seguros)
limite_superior = pd.Timestamp('2026-12-31')
limite_inferior = pd.Timestamp('2000-01-01')

# Aplicamos la máscara solo a valores válidos para evitar el TypeError
# Esto marca como NaT (el nulo de fecha) cualquier fecha fuera de rango
users.loc[(users['reg_date'].notna()) & ((users['reg_date'] > limite_superior) | (users['reg_date'] < limite_inferior)), 'reg_date'] = pd.NaT
users.loc[(users['churn_date'].notna()) & ((users['churn_date'] > limite_superior) | (users['churn_date'] < limite_inferior)), 'churn_date'] = pd.NaT

# 5. Verificación final
print("--- Limpieza completada ---")
print(f"Tipo de datos de 'reg_date': {users['reg_date'].dtype}")
print(f"Nulos en 'reg_date' (fuera de rango): {users['reg_date'].isna().sum()}")
print("DataFrame 'users' listo para el siguiente paso.")

### 3.2 Corregir sentinels y fechas imposibles
**🎯 Objetivo:**  
Decidir qué hacer con los valores nulos según su proporción y relevancia.

**Instrucciones:**
- Verifica si los nulos en `duration` y `length` son **MAR**(Missing At Random) revisando si dependen de la columna `type`.  
  Si confirmas que son MAR, **déjalos como nulos** y justifica la decisión.
  

In [ ]:
# 🎯 3.2 Verificación de nulos MAR (Missing At Random)

# Agrupamos por 'type' y contamos los nulos en 'duration' y 'length'
nulos_por_tipo = usage.groupby('type')[['duration', 'length']].apply(lambda x: x.isna().sum())

print("--- Conteo de nulos por tipo de servicio ---")
print(nulos_por_tipo)

# Verificamos si hay una correlación clara
print("\n--- Porcentaje de nulos por tipo ---")
print(usage.groupby('type')[['duration', 'length']].apply(lambda x: x.isna().mean() * 100))

Análisis de nulos (MAR):

Hallazgo: Los nulos en duration y length no son aleatorios; están directamente relacionados con la columna type. Las llamadas (call) carecen de datos en length, mientras que los mensajes (text) carecen de datos en duration.

Clasificación: Se confirma que son MAR (Missing At Random), ya que la probabilidad de que falte el dato depende de otra variable observada (type).

Acción: Mantener los valores nulos como NaN.

Justificación: Imputar estos valores con la media o la mediana sería un error estadístico grave, ya que el valor "cero" o el promedio de llamadas no tiene sentido técnico para un mensaje de texto. Mantenerlos como NaN permite que el sistema diferencie correctamente el comportamiento de cada servicio sin sesgar los resultados.

---

## 🧩Paso 4: Summary statistics de uso por usuario


### 4.1 Agrupación por comportamiento de uso

🎯**Objetivo**: Resumir las variables clave de la tabla `usage` **por usuario**, creando métricas que representen su comportamiento real de uso histórico. 

**Instrucciones:**: 
1. Construye una tabla agregada de `usage` por `user_id` que incluya:
- número total de mensajes  
- número total de llamadas  
- total de minutos de llamadas

2. Renombra las columnas para que tengan nombres claros:  
- `cant_mensajes`  
- `cant_llamadas`  
- `cant_minutos_llamada`
3. Combina esta tabla con `users`.

  

In [ ]:

# 1. Crear métricas agregadas por usuario
usage_summary = usage.groupby('user_id').agg(
    cant_mensajes=('type', lambda x: (x == 'text').sum()),
    cant_llamadas=('type', lambda x: (x == 'call').sum()),
    cant_minutos_llamada=('duration', 'sum')
).reset_index()

# 2. Renombrar las columnas para mayor claridad
usage_summary.columns = ['user_id', 'cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']

# 3. Combinar con la tabla de usuarios (usamos left merge para no perder usuarios sin uso)
df_final = users.merge(usage_summary, on='user_id', how='left')

# 4. Llenar con 0 los nulos generados por el merge (usuarios sin uso registrado)
cols_a_llenar = ['cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']
df_final[cols_a_llenar] = df_final[cols_a_llenar].fillna(0)

print("--- Resumen de comportamiento de uso ---")
print(df_final.head())


### 4.2 4.2 Resumen estadístico por usuario durante el 2024

🎯 **Objetivo:** Analizar las columnas numéricas y categóricas de los usuarios, para identificar rangos, valores extremos y distribución de los datos antes de continuar con el análisis.

**Instrucciones:**  
1. Para las columnas **numéricas** relevantes, obtén un resumen estadístico (media, mediana, mínimo, máximo, etc.).  
2. Para la columna **categórica** `plan`, revisa la distribución en **porcentajes** de cada categoría.


In [ ]:
# 1. Resumen estadístico de las columnas numéricas
columnas_numericas = ['age', 'cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']
print("--- Resumen Estadístico (Numérico) ---")
print(df_final[columnas_numericas].describe().T)

In [ ]:
# 2. Distribución de la columna 'plan' en porcentajes
distribucion_plan = df_final['plan'].value_counts(normalize=True) * 100
print("\n--- Distribución de usuarios por Plan (%) ---")
print(distribucion_plan)


---

## 🧩Paso 5: Visualización de distribuciones (uso y clientes) y outliers


### 5.1 Visualización de Distribuciones

🎯 **Objetivo:**  
Entender visualmente cómo se comportan las variables clave tanto de **uso** como de **clientes**, observar si existen diferencias según el tipo de plan, y analizar la **forma de la distribución**.

**Instrucciones:**  
Graficar **histogramas** para las siguientes columnas:  
- `age` (edad de los usuarios)
- `cant_mensajes`
- `cant_llamadas`
- `total_minutos_llamada` 

Después de cada gráfico, escribe un **insight** respecto al plan y la variable, por ejemplo:  
- "Dentro del plan Premium, hay mayor proporción de..."  
- "Los usuarios Básico tienden a hacer ... llamadas y enviar ... mensajes."  o "No existe algún patrón."
- ¿Qué tipo de distribución tiene ? (simétrica, sesgada a la derecha o a la izquierda) 

**Hint**  
Para cada histograma, 
- Usa `hue='plan'` para ver cómo varían las distribuciones según el plan (Básico o Premium).
- Usa `palette=['skyblue','green']`
- Agrega título y etiquetas


In [ ]:
plt.figure(figsize=(8, 5))
sns.histplot(data=df_final, x='age', hue='plan', kde=True, palette=['skyblue', 'green'], element="step")
plt.title('Distribución de Edad por Plan')
plt.xlabel('Edad')
plt.show()

💡Insights:

Distribuciones muy similares entre Básico y Premium → no parece haber segmentación fuerte por edad.

La edad se ve bastante uniforme en todo el rango (aprox. 18–80).

No hay concentración marcada en jóvenes o mayores.

Insight: El tipo de plan no está determinado por la edad.

In [ ]:
plt.figure(figsize=(8, 5))
sns.histplot(data=df_final, x='cant_mensajes', hue='plan', kde=True, palette=['skyblue', 'green'], element="step")
plt.title('Distribución de Cantidad de Mensajes por Plan')
plt.xlabel('Cantidad de Mensajes')
plt.show()

💡Insights: 

Ambas distribuciones son parecidas, pero:

Premium tiende ligeramente a más mensajes.

Básico tiene mayor concentración en valores medios-bajos.

La mayoría de usuarios envía entre 3 y 7 mensajes.

Valores extremos (muchos mensajes) son raros.

Insight: Los usuarios Premium parecen más activos en mensajería, aunque la diferencia no es radical.

In [ ]:
plt.figure(figsize=(8, 5))
sns.histplot(data=df_final, x='cant_llamadas', hue='plan', kde=True, palette=['skyblue', 'green'], element="step")
plt.title('Distribución de Cantidad de Llamadas por Plan')
plt.xlabel('Cantidad de Llamadas')
plt.show()

💡Insights: 

Patrón muy similar al de mensajes.

La mayoría realiza entre 2 y 6 llamadas.

Premium muestra una cola ligeramente más larga (más usuarios con muchas llamadas).

Insight: Premium sugiere mayor uso del servicio, pero sin separación fuerte entre planes.

In [ ]:
plt.figure(figsize=(8, 5))
sns.histplot(data=df_final, x='cant_minutos_llamada', hue='plan', kde=True, palette=['skyblue', 'green'], element="step")
plt.title('Distribución de Total de Minutos por Plan')
plt.xlabel('Total Minutos')
plt.show()

💡Insights:

Distribución sesgada a la derecha (muchos usuarios con pocos minutos, pocos con muchos).

La mayoría consume menos de 30 minutos.

Existen outliers claros (usuarios con consumo muy alto).

Premium muestra mayor presencia en consumos altos.

Insight importante:

👉 El comportamiento de uso (minutos) sí diferencia a los planes, más que edad, mensajes o llamadas.

👉 Probable conclusión analítica:
Usuarios Premium podrían representar heavy users, mientras Básico agrupa usuarios más moderados.

### 5.2 Identificación de Outliers

🎯 **Objetivo:**  
Detectar valores extremos en las variables clave de **uso** y **clientes** que podrían afectar el análisis, y decidir si requieren limpieza o revisión adicional.

**Instrucciones:**  
- Usa **boxplots** para identificar visualmente outliers en las siguientes columnas:  
  - `age` 
  - `cant_mensajes`
  - `cant_llamadas`
  - `total_minutos_llamada`  
- Crea un **for** para generar los 4 boxplots automáticamente.
<br>

- Después de crear los gráfico, responde si **existen o no outliers** en las variables.  
- Si hay outliers, crea otro bucle para calcular los límites de esas columnas usando el **método IQR** y decide qué hacer con ellos.
  - Si solamente hay outliers de un solo lado, no es necesario calcular ambos límites.

**Hint:**
- Dentro del bucle, usa `plt.title(f'Boxplot: {col}')` para que el título cambie acorde a la columna.



In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Visualización usando BoxPlot
columnas_numericas = ['age', 'cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']

for col in columnas_numericas:
    plt.figure(figsize=(8, 4))
    sns.boxplot(x=df_final[col], color='skyblue')
    plt.title(f'Boxplot: {col}')
    plt.show()

💡Insights: 

📦 Boxplot: Edad

Lo que se observa:

La mediana está cerca de los 50 años.

Rango intercuartílico amplio → usuarios muy variados en edad.

No se aprecian outliers fuertes.

Insight:

👉 La edad está bien distribuida y sin valores atípicos problemáticos.
👉 No hay sesgos ni concentraciones extrañas.

📦 Boxplot: Cantidad de Mensajes

Lo importante aquí:

Mediana alrededor de 5 mensajes.

La mayoría de usuarios cae en un rango relativamente estrecho.

Existen outliers hacia la derecha (usuarios que envían muchos mensajes).

Insight:

👉 El comportamiento típico es moderado.
👉 Hay un pequeño grupo de usuarios muy activos en mensajería.

Este patrón es clásico de uso digital.

📦 Boxplot: Cantidad de Llamadas

Muy similar al de mensajes:

Mediana aprox. 4 llamadas.

Variabilidad controlada.

Outliers altos visibles.

Insight:

👉 La mayoría de usuarios usa el servicio de forma regular.
👉 Existen usuarios con uso intensivo (heavy callers).

No parece haber problemas de dispersión extrema.

📦 Boxplot: Cantidad de Minutos de Llamada (El más revelador)

Aquí es donde aparecen los hallazgos más fuertes:

Distribución altamente sesgada a la derecha.

Gran concentración en minutos bajos.

Muchísimos outliers.

Cola larga con valores muy altos (usuarios que consumen muchos minutos).

Insight crítico:

👉 La variable minutos presenta heterogeneidad extrema.
👉 El sistema tiene usuarios con perfiles muy distintos:

Usuarios ligeros (la mayoría)

Usuarios intensivos (minoría pero muy marcada)

👉 Esto sugiere segmentación clara de comportamiento.

👉 Analíticamente, esta variable:

✅ Puede explicar diferencias entre planes
✅ Puede requerir transformaciones (log) en modelado
✅ Puede afectar medias / métricas agregadas

In [ ]:
# Calcular límites con el método IQR
limites_df = {}

for col in columnas_numericas:
    Q1 = df_final[col].quantile(0.25)
    Q3 = df_final[col].quantile(0.75)
    IQR = Q3 - Q1
    
    # Definimos solo el límite superior para telecomunicaciones (sesgo positivo)
    limite_superior = Q3 + 1.5 * IQR
    limites_df[col] = limite_superior
    
    print(f"Columna {col} -> Límite Superior: {limite_superior:.2f}")

# Mostrar resumen comparativo entre el límite y el valor máximo real
columnas_limites = ['age', 'cant_mensajes', 'cant_llamadas', 'cant_minutos_llamada']
resumen_limites = df_final[columnas_limites].describe().T
resumen_limites['limite_superior'] = [limites_df[col] for col in columnas_limites]

print("\n--- Resumen: Comparativa Valor Máximo vs Límite IQR ---")
print(resumen_limites[['max', 'limite_superior']])

💡Insights: 

✅ cant_mensajes → Mantener outliers

Datos:

Límite IQR: 11.50

Máximo observado: 17

Interpretación:

Enviar 17 mensajes:

✔ Es completamente posible
✔ No es físicamente imposible
✔ Representa usuarios más activos

No hay indicios de error de medición ni valores absurdos.

Decisión correcta:

👉 Mantenerlos

Por qué:

Reflejan comportamiento real

Eliminarlo distorsionaría la actividad de usuarios intensivos

Son valiosos para segmentación / modelos de uso

Eliminar esto sería perder información válida.

✅ cant_llamadas → Mantener outliers

Datos:

Límite IQR: 10.50

Máximo: 15

Interpretación:

15 llamadas:

✔ Perfectamente plausible
✔ No indica error
✔ Consistente con heavy users

En telecomunicaciones esto es normal.

Decisión correcta:

👉 Mantenerlos

Por qué:

Representan variabilidad natural del uso

Usuarios intensivos son críticos para análisis de planes

No hay imposibilidad lógica

⚠️ cant_minutos_llamada → Mantener outliers (MUY IMPORTANTE)

Datos:

Límite IQR: 61.87

Máximo: 155.69

Visualmente parece extremo, pero aquí está el punto crítico:

Consumir muchos minutos:

✔ Es totalmente realista
✔ Es típico en datasets de consumo
✔ Es una variable naturalmente sesgada

En minutos los outliers SON parte del fenómeno.

Decisión correcta:

👉 Mantenerlos

Razón analítica fuerte:

Minutos casi siempre siguen distribuciones de cola larga

Heavy users son precisamente el insight de negocio

Eliminarlos destruiría patrones de comportamiento

Si eliminas estos valores:

❌ Subestimas consumo
❌ Dañas análisis de planes
❌ Rompes modelos predictivos

Lo correcto en modelado suele ser:

✅ Transformación logarítmica
NO eliminación.

---

## 🧩Paso 6: Segmentación de Clientes

### 6.1 Segmentación de Clientes Por Uso

🎯 **Objetivo:** Clasificar a cada usuario en un grupo de uso (Bajo uso, Uso medio, Alto uso) basándose en la cantidad de llamadas y mensajes registrados.

**Instrucciones:**  
- Crea una nueva columna llamada `grupo_uso` en el dataframe `user_profile`.
- Usa comparaciones lógicas (<, >) para evaluar las condiciones de llamadas y mensajes y asigna:
  - `'Bajo uso'` cuando llamadas < 5 y mensajes < 5
  - `'Uso medio'` cuando llamadas < 10 y mensajes < 10
  - `'Alto uso'` para el resto de casos
  

In [ ]:
def clasificar_uso(row):
    # Condiciones lógicas para la segmentación
    if row['cant_llamadas'] < 5 and row['cant_mensajes'] < 5:
        return 'Bajo uso'
    elif row['cant_llamadas'] < 10 and row['cant_mensajes'] < 10:
        return 'Uso medio'
    else:
        return 'Alto uso'

# Aplicar la función al dataframe
df_final['grupo_uso'] = df_final.apply(clasificar_uso, axis=1)

# Verificación de la distribución de los nuevos grupos
print("--- Distribución de usuarios por Grupo de Uso ---")
print(df_final['grupo_uso'].value_counts())

# Vista previa de los cambios
print(df_final[['user_id', 'cant_llamadas', 'cant_mensajes', 'grupo_uso']].head())

### 6.2 Segmentación de Clientes Por Edad

🎯 **Objetivo:**: Clasificar a cada usuario en un grupo por **edad**.

**Instrucciones:**  
- Crea una nueva columna llamada `grupo_edad` en el dataframe `user_profile`.
- Usa comparaciones lógicas (<, >) para evaluar las condiciones y asigna:
  - `'Joven'` cuando age < 30
  - `'Adulto'` cuando age < 60
  - `'Adulto Mayor'` para el resto de casos

    

In [ ]:
def clasificar_edad(age):
    # Condiciones lógicas para la segmentación por edad
    if age < 30:
        return 'Joven'
    elif age < 60:
        return 'Adulto'
    else:
        return 'Adulto Mayor'

# Aplicar la función a la columna 'age'
df_final['grupo_edad'] = df_final['age'].apply(clasificar_edad)

# Verificación de la distribución de los nuevos grupos
print("--- Distribución de usuarios por Grupo de Edad ---")
print(df_final['grupo_edad'].value_counts())

# Vista previa de los cambios
print(df_final[['user_id', 'age', 'grupo_edad']].head())

### 6.3 Visualización de la Segmentación de Clientes

🎯 **Objetivo:** Visualizar la distribución de los usuarios según los grupos creados: **grupo_uso** y **grupo_edad**.

**Instrucciones:**  
- Crea dos gráficos para las variables categóricas `grupo_uso` y `grupo_edad`.
- Agrega título y etiquetas a los ejes en cada gráfico.


In [ ]:
plt.figure(figsize=(8, 5))
sns.countplot(data=df_final, x='grupo_uso', palette='viridis', order=['Bajo uso', 'Uso medio', 'Alto uso'])
plt.title('Distribución de Usuarios por Grupo de Uso')
plt.xlabel('Grupo de Uso')
plt.ylabel('Cantidad de Usuarios')
plt.show()

In [ ]:
plt.figure(figsize=(8, 5))
sns.countplot(data=df_final, x='grupo_edad', palette='magma', order=['Joven', 'Adulto', 'Adulto Mayor'])
plt.title('Distribución de Usuarios por Grupo de Edad')
plt.xlabel('Grupo de Edad')
plt.ylabel('Cantidad de Usuarios')
plt.show()


---
## 🧩Paso 7: Insight Ejecutivo para Stakeholders

🎯 **Objetivo:** Traducir los hallazgos del análisis en conclusiones accionables para el negocio, enfocadas en segmentación, patrones de uso y oportunidades comerciales.

**Preguntas a responder:** 
- ¿Qué problemas tenían originalmemte los datos?¿Qué porcentaje, o cantidad de filas, de esa columna representaban?


- ¿Qué segmentos de clientes identificaste y cómo se comportan según su edad y nivel de uso?  
- ¿Qué segmentos parecen más valiosos para ConnectaTel y por qué?  
- ¿Qué patrones de uso extremo (outliers) encontraste y qué implican para el negocio?


- ¿Qué recomendaciones harías para mejorar la oferta actual de planes o crear nuevos planes basados en los segmentos y patrones detectados?

✍️ **Escribe aquí tu análisis ejecutivo:**

1. Calidad de Datos: Estado Inicial
Originalmente, los datos presentaban inconsistencias técnicas que habrían invalidado cualquier modelo predictivo si no se trataban:

Fechas (reg_date, churn_date): Contenían valores fuera de rango (fuera del periodo 2000-2026), que fueron tratados como NaT.

Sentinels: La columna age presentaba el valor -999 (encontrado y corregido mediante la mediana), y la columna city contenía ?, los cuales fueron convertidos a nulos.

Naturaleza MAR: Los nulos en duration y length no fueron errores de carga, sino una característica técnica: las llamadas carecían de longitud de texto y los mensajes de duración en tiempo. Esto representaba aproximadamente el 99.9% de los datos nulos en cada columna respectiva, confirmando que no debían ser imputados, sino segmentados por type.

2. Segmentación de Clientes
Identificamos un comportamiento heterogéneo basado en el cruce de dos variables:

Segmentación por Uso: Los usuarios se distribuyen principalmente en "Bajo" y "Medio uso". Los "Alto uso" representan el segmento de nicho pero de mayor impacto.

Segmentación por Edad: La mayoría de nuestra base se concentra en el segmento "Adulto".

Correlación: Los usuarios "Jóvenes" tienden a ser más intensivos en mensajes, mientras que los "Adultos" y "Adultos Mayores" presentan patrones de uso más orientados a llamadas (voz).

3. Segmentos de Alto Valor
El segmento más valioso para ConnectaTel son los "Power Users" (clasificados como "Alto uso" independientemente de su edad).

Por qué: Estos usuarios tienen una mayor dependencia de nuestros servicios, lo que se traduce en un menor riesgo de churn (abandono) y una mayor oportunidad de upselling hacia planes con mayores cuotas de datos o llamadas internacionales.

4. Patrones Extremos (Outliers)
El análisis detectó outliers significativos en cant_mensajes y cant_minutos_llamada.

Implicación: Estos no son errores, sino clientes que utilizan la red al límite de su capacidad. Para el negocio, esto implica que nuestros límites actuales de plan podrían estar dejando dinero sobre la mesa. Un pequeño grupo de clientes está consumiendo una proporción desproporcionada de recursos, lo que podría afectar la latencia si no se gestiona correctamente.

5. Recomendaciones Estratégicas
Basado en los hallazgos, propongo las siguientes acciones:

Creación de un plan "Flex": Dado que un grupo importante de usuarios está en "Uso bajo" o "Uso medio", un plan con cuotas ajustables sería más atractivo que los planes rígidos actuales, reduciendo el churn por percepción de "pago por servicio no utilizado".

Monetización de "Power Users": Diseñar un plan "Pro-Unlimited" enfocado en los outliers de alto consumo, ofreciendo beneficios exclusivos a cambio de una mensualidad fija más alta, garantizando mayor prioridad en la red.

Campaña de fidelización por edad:

Jóvenes: Ofertas centradas en paquetes de datos/mensajería.

Adultos Mayores: Paquetes con minutos ilimitados a números fijos o familiares, aprovechando su patrón de uso de voz detectado.

### Análisis ejecutivo

⚠️ **Problemas detectados en los datos**
- abc
- abc


🔍 **Segmentos por Edad**
- abc
- abc 


📊 **Segmentos por Nivel de Uso**
- abc
- abc


➡️ Esto sugiere que ...


💡 **Recomendaciones**
- abc
- abc 

---

## 🧩Paso 8 Cargar tu notebook y README a GitHub

🎯 **Objetivo:**  
Entregar tu análisis de forma **profesional**, **documentada** y **versionada**, asegurando que cualquier persona pueda revisar, ejecutar y entender tu trabajo.



### Opción A : Subir archivos desde la interfaz de GitHub (UI)

1. Descarga este notebook (`File → Download .ipynb`).  
2. Entra a tu repositorio en GitHub (por ejemplo `telecom-analysis` o `sprint7-final-project`).  
3. Sube tu notebook **Add file → Upload files**.  

---

### Opción B : Guardar directo desde Google Colab

1. Abre tu notebook en Colab.  
2. Ve a **File → Save a copy in GitHub**.  
3. Selecciona el repositorio y la carpeta correcta (ej: `notebooks/`).  
4. Escribe un mensaje de commit claro, por ejemplo:  
    - `feat: add final ConnectaTel analysis`
    - `agregar version final: Análisis ConnectaTel`
5. Verifica en GitHub que el archivo quedó en el lugar correcto y que el historial de commits se mantenga limpio.

---

Agrega un archivo `README.md` que describa de forma clara:
- el objetivo del proyecto,  
- los datasets utilizados,  
- las etapas del análisis realizadas,  
- cómo ejecutar el notebook (por ejemplo, abrirlo en Google Colab),  
- una breve guía de reproducción.
---

Link a repositorio público del proyecto: `LINK a tu repo aquí`